src/classes/RD_RecurringDonations.cls

```diff
@@ -795,77 +766,391 @@ public class RD_RecurringDonations {
+
+        for (npe03__Recurring_Donation__c r : recurringDonations) {
+
+            Date nextDate = getStartDate(r);
+
+            integer rdcounter = 1;
+            if (r.npe03__Total_Paid_Installments__c > 0) {
+                rdcounter = (integer)r.npe03__Total_Paid_Installments__c + 1;
+            }
+
+            // loop the existing opportunities
+            for (Opportunity o : RDOppMap.get(r.Id)) {
+                if (nextDate <= dtForecast) {
+                    // fix close dates on existing opportunities
+                    o.CloseDate = nextDate;
+
+                    // TODO: are these really all the fields necessary to sync to oppties on an update?
+                    // This is borrowed from roughly line 700
+                    o.Amount = r.npe03__Amount__c;
+                    if (r.npe03__Recurring_Donation_Campaign__c != null && rds.npe03__Add_Campaign_to_All_Opportunites__c) {
+                        o.CampaignId = r.npe03__Recurring_Donation_Campaign__c;
+                    }
+
+                    if (r.npe03__Organization__c != null) {
+                        o.AccountID = r.npe03__Organization__c;
+                    }
+
+                    //modify the name
+                    String oName = '';
+                    oName += r.npe03__Donor_Name__c;
+                    oName += ' ';
+                    oName += system.label.npe03.RecurringDonationPrefix;
+                    oName += ' (' + o.Recurring_Donation_Installment_Number__c + ') ';
+                    oName += o.CloseDate.format();
+                    o.Name = oName;
+
+                    //evaluate any custom field mappings
+                    for (string s : customFieldMappingSettings.keySet()) {
+                        npe03__Custom_Field_Mapping__c cfm = customFieldMappingSettings.get(s);
+                        o.put(cfm.npe03__Opportunity_Field__c, r.get(cfm.npe03__Recurring_Donation_Field__c));
+                    }
+
+                    updateOppsList.add(o);
+
+                    nextDate = getNextDate(nextDate, r);
+                    rdcounter++;
+
+                } else {
+                    // extras get closed
+                    closeOppsList.add(o);
+                }
+            }
+
+            // didn't have enough, create more
+            while (nextDate < dtForecast) {
+                Opportunity o = newOpportunityOnDate(nextDate,r,rdcounter);
+                insertOppsList.add(o);
+                nextDate = getNextDate(nextDate, r);
+                rdcounter++;
+            }
+        }
+
+        if (!insertOppsList.isEmpty()) {
+            Database.SaveResult[] lsr;
+            list<Database.SaveResult> badSRs = new list<Database.SaveResult>();
+            list<Opportunity> badOpps = new list<Opportunity>();
+
+            lsr = Database.insert(insertOppsList, false);
+
+            set<id> updateRecurringDonationSet = new set<id>();
+            for (Opportunity o : insertOppsList) {
```

Comment:
Not sure I'm clear on what happens if there are insert errors here because the code to check for errors is below this loop.